### PR TITLE
Add support for PangoAttribute

### PR DESCRIPTION
**Pull request by naveen521kk**
Planning to add API support for PangoAttributes. ~This is a draft PR which requires some amount of work.~
Done needs review.

### CODE REVIEW

File: pangocffi/attributes.py [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+        """
+        API not implemented
+        """
+        temp = cls.from_pointer(pango.pango_attr_language_new(language))
+        temp.start_index = start_index
+        temp.end_index = end_index
+        return temp
+
+    @classmethod
+    def from_family(
```
Comment (naveen521kk) [inline pangocffi/attributes.py:94]: @leifgehrmann can you suggest me a better way that this?

**Comment by naveen521kk** [comment]
Without increasing the `ref()` in `AttrList` Python crashes what to do @leifgehrmann ?

**Comment by naveen521kk** [comment]
Added tests and everything got covered. 
Needs your review. @leifgehrmann

**Comment by naveen521kk** [comment]
I noticed some warnings from glib when running the test. Do you why @leifgehrmann ?

**Comment by naveen521kk** [comment]
@leifgehrmann can you review it?

**Review by leifgehrmann — CHANGES_REQUESTED** [review]
Hi, I've left a few suggestions and fixes, but generally this looks great! Apologies for the delay in review...

If you prefer, I could also just merge in the changes now and apply the suggestions myself. The main structural change I suggested was splitting `attributes.py` into two files.

Regarding some of the questions above:

> Without increasing the ref() in AttrList Python crashes what to do @leifgehrmann ?

Is this issue resolved after you posted that comment? I see we call `self._ref()` in `insert`, `insert_before` and `change`.

To be honest, I'm not sure why calling `self._ref()` is necessary since pango's own tests don't need to explicitly call `pango_attr_list_ref`. I would think so worst case, we might be causing memory leaks because we are not calling an `self._unref()` somewhere... but if that's really an issue we can resolve it later I think.

> I noticed some warnings from glib when running the test. Do you why @leifgehrmann ?

I looked in the logs of the Ubuntu, MacOS and Windows checks and couldn't find any warnings (except for the `Warning: glib 2.66.2_1 is already installed and up-to-date` in the install stage of the MacOS one). What glib warnings are you referring to?

Or do you mean the warnings in the Ubuntu checks? That will probably be because GitHub considers "Ubuntu-Latest" to be 18.04.5. This version of Ubuntu doesn't have the latest version of `libpango` it seems, which is why those warnings appear... But that's not a huge issue I think.

Or do you mean something else entirely? 😅

**Comment by naveen521kk** [comment]
Good to hear from you :)

> If you prefer, I could also just merge in the changes now and apply the suggestions myself. The main structural change I suggested was splitting attributes.py into two files.

You can. I will be able to access my workspace only tomorrow if you don't want to wait then you can proceed.

> Is this issue resolved after you posted that comment? I see we call self._ref() in insert, insert_before and change.

I pretty much had to call `self._ref()` only then it was resolved. I really don't know why it happens here so. I was playing around with Cython and Pango, and didn't face this issue there also.

> I would think so worst case, we might be causing memory leaks because we are not calling an self._unref() somewhere... but if that's really an issue we can resolve it later I think.

Could be I don't really know much about this.

> What glib warnings are you referring to?

This happened in my local environment. 
image
this came again and again for each test. See https://pastebin.com/HR85qSBj

File: pangocffi/attributes.py [hunk 2]
```diff
@@ -0,0 +1,1085 @@
...
+        Set's the End Index of a Attribute.
+
+        :param end_index:
+            end index of the range. The character at this index
+            is not included in the range.
+        :raises: AssetionError
+            When ``end_index`` isn't a :class:`int`.
+        """
+        assert isinstance(end_index, int), "end_index isn't a int"
+        # assert end_index >= 0
```
[the remark this answers is not in the document]
Comment (naveen521kk) [reply pangocffi/attributes.py:91]: Adding an assert statement would do good :)

File: pangocffi/attributes.py (hunk 1 above)
Comment (naveen521kk) [reply pangocffi/attributes.py:109]: :+1: to this.

**Comment by naveen521kk** [comment]
I just tried applying suggestions. Also, If you want you can fix other things now, or I will do tomorrow.

File: pangocffi/attributes.py (hunk 1 above)
Comment (naveen521kk) [reply pangocffi/attributes.py:109]: Note: I copied the end_index from https://developer.gnome.org/pygtk/stable/class-pangoattribute.html idk why they defined so. But there could be a reason for it.

File: pangocffi/layout.py [hunk 3]
```diff
@@ -340,3 +339,23 @@ def get_iter(self) -> LayoutIter:
...
+        """
+        Sets the text attributes for a layout object.
+
+        :param attrs: a :class:`AttrList`
+        :type attrs: AttrList
+        """
+        if attrs is None:
+            pango.pango_layout_set_attributes(self._pointer, ffi.NULL)
+        else:
+            pango.pango_layout_set_attributes(self._pointer, attrs.get_pointer())
```
Comment (leifgehrmann) [inline pangocffi/layout.py:353]:
```suggestion
            pango.pango_layout_set_attributes(
                self._pointer,
                attrs.get_pointer()
            )
```

File: pangocffi/attributes.py [hunk 4]
```diff
@@ -0,0 +1,1085 @@
...
+            the start index of the range. Should be >=0.
+        :param end_index:
+            end index of the range. The character at this index
+            is not included in the range.
+        :return:
+            the Attribute.
+        :raises: AssetionError
+            When ``absolute_size`` isn't a :class:`int`.
+        """
+        assert isinstance(absolute_size, int), "absolute_size isn't int"
```
Comment (leifgehrmann) [inline pangocffi/attributes.py:277]:
```suggestion
        assert isinstance(size, int), "size isn't int"
```

File: pangocffi/attributes.py [hunk 5]
```diff
@@ -0,0 +1,1085 @@
...
+        absolute_size = ffi.cast("int", absolute_size)
```
Comment (leifgehrmann) [inline pangocffi/attributes.py:278]:
```suggestion
        size = ffi.cast("int", size)
```

File: pangocffi/attributes.py [hunk 6]
```diff
@@ -0,0 +1,1085 @@
...
+        temp = cls._init_pointer(
+            pango.pango_attr_size_new_absolute(
+                absolute_size,
```
Comment (leifgehrmann) [inline pangocffi/attributes.py:281]:
```suggestion
                size,
```

File: pangocffi/attributes.py [hunk 7]
```diff
@@ -0,0 +1,1085 @@
...
+        assert isinstance(attr_list, AttrList), "attr_list isn't a AttrList"
+        assert isinstance(pos, int), "pos isn't a int"
+        assert isinstance(length, int), "length isn't a int"
+        length = ffi.cast("gint", length)
+        pos = ffi.cast("gint", pos)
+        pango.pango_attr_list_splice(
+            self._pointer,
+            attr_list._pointer,
+            pos,
+            leng,
```
Comment (leifgehrmann) [inline pangocffi/attributes.py:1072]:
```suggestion
            length,
```

**Review by leifgehrmann — COMMENTED** [review]
Just some linting fixes to get the PR merged

File: pangocffi/attributes.py (hunk 1 above)
Comment (leifgehrmann) [reply pangocffi/attributes.py:109]: Thanks for referencing link, I hadn't seen it!

In that case, I think it's more reasonable for it to remain as 1 then. I don't want to diverge too much from their API and cause confusion.

**Comment by leifgehrmann** [comment]
> This happened in my local environment.
> image
> this came again and again for each test. See https://pastebin.com/HR85qSBj

Interesting... It's surprising that this is happening on your machine, but not in the checks (Unless GitHub Actions are suppressing these warning somehow)

<summary></summary>
<details>Screenshot 2020-12-18 at 22 53 45</details>

Are your DLLs for glib the same as the DLL being used in the GitHub Action?

**Comment by naveen521kk** [comment]
> Unless GitHub Actions are suppressing these warning somehow

Definitely, Github Action suppresses it. 

> Are your DLLs for glib the same as the DLL being used in the GitHub Action?

I use the same one I build.
Also, I tried the same on msys2, where I could see the warnings only when I run `pytest -s` and not simply pytest.

**Comment by leifgehrmann** [comment]
> Also, I tried the same on msys2, where I could see the warnings only when I run `pytest -s` and not simply pytest.

Excellent, I can easily replicate this even on MacOS. I've raised this issue in #35.

**Pull request merged**